### PR TITLE
Update Omitted Type to use Hashable Values as Keys for Caching

### LINKED ISSUE #6957: Caching of functions with keyword arguments
## Reporting a bug

- [X] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [X] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

Using keyword arguments somehow results in the cache being recreated for each process.

Using the following script:

```python
from numba import njit
import numpy as np


@njit(cache=True)
def example(x, y, eps=1e-12):
    result = x + y
    result[result < eps] = 0
    return result


if __name__ == '__main__':
    rng = np.random.default_rng(0)
    x = rng.normal(0, 0.1, 50)
    y = rng.normal(0, 0.1, len(x))
    print(example(x, y))
```

Results in this behaviour when called multiple times:

```
❯ NUMBA_DEBUG_CACHE=1 python fitting.py  | grep saved
[cache] index saved to '__pycache__/fitting.example-5.py38.nbi'
[cache] data saved to '__pycache__/fitting.example-5.py38.1.nbc'
❯ NUMBA_DEBUG_CACHE=1 python fitting.py  | grep saved
[cache] index saved to '__pycache__/fitting.example-5.py38.nbi'
[cache] data saved to '__pycache__/fitting.example-5.py38.2.nbc'
❯ NUMBA_DEBUG_CACHE=1 python fitting.py  | grep saved
[cache] index saved to '__pycache__/fitting.example-5.py38.nbi'
[cache] data saved to '__pycache__/fitting.example-5.py38.3.nbc'
```

Observe that the cache is written each time, with an increasing number in the filename. This created problems when running a numba function on a cluster, generating multiple GB of caching data.

Removing the keyword argument solves the issue:

```
from numba import njit
import numpy as np


@njit(cache=True)
def example(x, y):
    eps = 1e-12
    result = x + y
    result[result < eps] = 0
    return result


if __name__ == '__main__':
    rng = np.random.default_rng(0)
    x = rng.normal(0, 0.1, 50)
    y = rng.normal(0, 0.1, len(x))
    print(example(x, y))

```

Observe no "saved" output on second and third run:

```
❯ NUMBA_DEBUG_CACHE=1 python fitting.py  | grep saved
[cache] index saved to '__pycache__/fitting.example-5.py38.nbi'
[cache] data saved to '__pycache__/fitting.example-5.py38.1.nbc'
❯ NUMBA_DEBUG_CACHE=1 python fitting.py  | grep saved
❯ NUMBA_DEBUG_CACHE=1 python fitting.py  | grep saved
```

### PR DESCRIPTION
**Pull request by njriasan**
Closes #6957.

Changes Omitted arguments to use a different key for hashable and non-hashable types. For hashable types, this enables loading cache from disk even if id no longer matches.

### CODE REVIEW

**Comment by esc** [comment]
@njriasan thank you very much for submitting this. I have queued it for a review by one of the devs.

**Comment by njriasan** [comment]
@sklam Thanks for the feedback! I restructured the code based on your suggestions.

File: numba/tests/test_caching.py [hunk 1]
```diff
@@ -43,3 +47,71 @@ def test_constant_unicode_cache(self):
...
+                "Cache has an unexpected hit"
+        success = True
+    # Catch anything raised so it can be propagated
+    except: # noqa: E722
+        output = traceback.format_exc()
+        success = False
+    result_queue.put((success, output))
+
+
+class TestDiskCaching(SerialMixin, TestCase):
```
[the remark this answers is not in the document]
Comment (njriasan) [reply numba/tests/test_caching.py]: Agreed.

File: numba/tests/test_caching.py [hunk 2]
```diff
@@ -43,3 +47,71 @@ def test_constant_unicode_cache(self):
...
+    """
+        Tests that caching works when loadingvalues from disk.
+    """
+    def test_omitted(self):
+
+        # Test in a new directory
+        cache_dir = temp_directory(self.__class__.__name__)
+        with override_config("CACHE_DIR", cache_dir):
```
[the remark this answers is not in the document]
Comment (njriasan) [reply numba/tests/test_caching.py]: Thank you. I have fixed it so it works.

File: numba/core/types/misc.py [hunk 3]
```diff
@@ -87,11 +88,19 @@ class Omitted(Opaque):
 
     def __init__(self, value):
         self._value = value
+        # Because id(value) may not match when loading from disk,
+        # just return the value if we can hash it.
+        # See discussion in gh #6957
+        try:
+            hash(value)
+            self._value_key = value
```
[the remark this answers is not in the document]
Comment (njriasan) [reply numba/core/types/misc.py]: Done.

**Comment by njriasan** [comment]
@sklam Can we try get this included in 0.55? I think this is close to being ready.

**Review by sklam — APPROVED** [review]
Thanks for the patch!

**Pull request merged**